### PR TITLE
Support different LLMs besides OpenAI via Langchain

### PR DESCRIPTION
**Pull request by finger-bone**
This PR addresses the issue #26 .

This PR made the `LLMClient` an abstract class, and renamed the original `LLMClient` to `OpenAILLMClient`, with no code change, then implemented LcLLMClient as Langchain LLM Client, which takes any Langchain ChatModel and adapts it into the code. In addition, `BaseParser` had `get_api_key` call from `LLMClient` removed. This call was not used in the old code.

`extract_entities_from_pdf_with_langchain_client.py` is added for testing. The only difference from the `extract_entities_from_pdf.py` is that the new script uses LcLLMClient.

### CODE REVIEW

File: pyproject.toml
Comment (VinciGit00) [inline pyproject.toml:1]: Hi, I do not see any langchain import here, please can you add it?

File: scrapontology/llm_client/openai_llm_client.py
Comment (VinciGit00) [inline scrapontology/llm_client/openai_llm_client.py:1]: Please instead of using requests can you use chatOpenai? (https://python.langchain.com/docs/integrations/chat/openai/) 
Take a look on Scrapegraphai how models are iniated.
You should take a look on these examples from scrapegraphai: https://github.com/ScrapeGraphAI/Scrapegraph-ai/blob/main/scrapegraphai/graphs/abstract_graph.py

File: examples/extract_entities_json_schema_from_pdf.py
Comment (VinciGit00) [inline examples/extract_entities_json_schema_from_pdf.py:1]: I would like an initialization of the examples like this please https://github.com/ScrapeGraphAI/Scrapegraph-ai/blob/main/examples/openai/smart_scraper_openai.py

Comment (VinciGit00) [reply examples/extract_entities_json_schema_from_pdf.py:1]: (for all the examples)

**Review by VinciGit00 — CHANGES_REQUESTED** [review]

File: scrapontology/llm_client/openai_llm_client.py
Comment (finger-bone) [reply scrapontology/llm_client/openai_llm_client.py:1]: So, if I'm getting it right, you'd prefer using some parameters or parameter dictionary to initialize the LLMClient instead of just passing a langchain chatllm object (now), is that correct? If so, I'll just make corresponding changes.

In addition, if we were to implement a more general LLMClient that supports major langchain chat llms, then there will be no need to implement a more restricted class that only supports openai client, since you siad that you don't want the current requests-version and prefer langchain.

Comment (VinciGit00) [reply scrapontology/llm_client/openai_llm_client.py:1]: yes I would like to have something like that and te initialization is done under the hood by the library.
From our experience is easier in this way and more fault tollerant

Comment (finger-bone) [reply scrapontology/llm_client/openai_llm_client.py:1]: get it, the new commit has addressed that, check if this is how you wanted.

File: scrapontology/llm_client.py
Comment (VinciGit00) [inline scrapontology/llm_client.py:1]: Hi @finger-bone I think is almost perfect! You are in the rigth way.

I think the best solution should be integrate init_chat_model from langchain. Here's the documentation https://js.langchain.com/docs/how_to/chat_models_universal_init/.

In this way we can skip the vast amount of llm_models initialization  and is more update-tolerant.

Thank you for the effort you are putting for this library.

If you want please write me an email (mvincig11@gmail.com)

Comment (finger-bone) [reply scrapontology/llm_client.py:1]: sry i didn't know this feature before, thanks for the hints! Have done that.

Comment (VinciGit00) [reply scrapontology/llm_client.py:1]: @finger-bone one more thing.
Please can you update all the dependencies inside pyproject.toml with >= instead of ==?
Thank you again

Comment (finger-bone) [reply scrapontology/llm_client.py:1]: sure thing

File: pyproject.toml
Comment (VinciGit00) [inline pyproject.toml:1]: can you add also the support for ollama? After this is finished

**Comment by lurenss** [comment]
Hi @finger-bone, thank you very much, for your work.

One small thing is needed before merging.

Could you add support for ollama, always using langchain?

Thank you in advance

**Comment by finger-bone** [comment]
> Could you add support for ollama, always using langchain?

I suppose it already does? Just missing a depenedency, i've added that. 
image

**Review by VinciGit00 — APPROVED** [review]
OK thank you

**Comment by lurenss** [comment]
lgtm

**Pull request merged**
